lib/charms/nginx_ingress_integrator/v0/nginx_route.py

```diff
@@ -0,0 +1,383 @@
+            owasp-modsecurity-crs  option via relation, optional.
+        owasp_modsecurity_custom_rules: configure Nginx ingress
+            integrator owasp-modsecurity-custom-rules option via
+            relation, optional.
+        path_routes: configure Nginx ingress integrator path-routes
+            option via relation, optional.
+        retry_errors: configure Nginx ingress integrator retry-errors
+            option via relation, optional.
+        rewrite_target: configure Nginx ingress integrator
+            rewrite-target option via relation, optional.
+        rewrite_enabled: configure Nginx ingress integrator
+            rewrite-enabled option via relation, optional.
+        service_namespace: configure Nginx ingress integrator
+            service-namespace option via relation, optional.
+        session_cookie_max_age: configure Nginx ingress integrator
+            session-cookie-max-age option via relation, optional.
+        tls_secret_name: configure Nginx ingress integrator
+            tls-secret-name option via relation, optional.
+        nginx_route_relation_name: Specifies the relation name of
+            the relation handled by this requirer class. The relation
+            must have the nginx-route interface.
+    """
+    config: typing.Dict[str, typing.Union[str, int, bool]] = {}
+    if service_hostname is not None:
+        config["service-hostname"] = service_hostname
+    if service_name is not None:
+        config["service-name"] = service_name
+    if service_port is not None:
+        config["service-port"] = service_port
+    if additional_hostnames is not None:
+        config["additional-hostnames"] = additional_hostnames
+    if limit_rps is not None:
+        config["limit-rps"] = limit_rps
+    if limit_whitelist is not None:
+        config["limit-whitelist"] = limit_whitelist
+    if max_body_size is not None:
+        config["max-body-size"] = max_body_size
+    if owasp_modsecurity_crs is not None:
+        config["owasp-modsecurity-crs"] = owasp_modsecurity_crs
+    if owasp_modsecurity_custom_rules is not None:
+        config["owasp-modsecurity-custom-rules"] = owasp_modsecurity_custom_rules
+    if path_routes is not None:
+        config["path-routes"] = path_routes
+    if retry_errors is not None:
+        config["retry-errors"] = retry_errors
+    if rewrite_target is not None:
+        config["rewrite-target"] = rewrite_target
+    if rewrite_enabled is not None:
+        config["rewrite-enabled"] = rewrite_enabled
+    if service_namespace is not None:
+        config["service-namespace"] = service_namespace
+    if session_cookie_max_age is not None:
+        config["session-cookie-max-age"] = session_cookie_max_age
+    if tls_secret_name is not None:
+        config["tls-secret-name"] = tls_secret_name
+
+    _NginxRouteRequirer(
+        charm=charm, config=config, nginx_route_relation_name=nginx_route_relation_name
+    )
+
+
+class _NginxRouteProvider(Object):
```

Comment:
Is it necessary to inherit from `Object`?

Reply:
This is the `Object` imported from the `ops.framework`.

Reply:
Hmm, do you know what that class does? It is quite confusing given the `object` inbuilt type. Maybe we can give the `ops` team feedback on potentially improving the name so that it is clearer what that class does?

Reply:
It's the base class for all entities in the `ops.framework`. And yes, it's confusion.

Reply:
@benhoyt Any thoughts?

Reply:
Yeah, it's probably not a great name choice in `ops`. If you `from ops import framework` then you'd refer to it as `framework.Object`, which would be a bit clearer. In future we're [planning](https://github.com/canonical/operator/issues/731) to make it so all public names are accessible via just `import ops` so you'll be able to say `ops.Object`, but that's not available yet.

In any case, I believe you only need to inherit from `ops.framework.Object` if you're a "framework object" like an event that's going to need to be serialized (every `Object` has a unique path). In this case I don't think you need that -- it looks like the type *has* events but *is not* an event. (I haven't take a super-close look though.)

Reply:
Makes sense, `framework.Object` if it is needed or remove it otherwise. Perhaps worth updating the documentation of the class to explain when it is necessary? It isn't immediately obvious, at least to me, from the docstring:

```Python
class Object:
    """Initialize an Object as a new leaf in :class:`Framework`, identified by `key`.

    Args:
        parent: parent node in the tree.
        key: unique identifier for this object.

    Every object belongs to exactly one framework.

    Every object has a parent, which might be a framework.

    We track a "path to object," which is the path to the parent, plus the object's unique
    identifier. Event handlers use this identity to track the destination of their events, and the
    Framework uses this id to track persisted state between event executions.

    The Framework should raise an error if it ever detects that two objects with the same id have
    been created.

    """
```

Reply:
I will just use import to replace all import-from statements for now.